src/Util/ComposeFileManipulator.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/*
+ * This file is part of the Symfony MakerBundle package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\MakerBundle\Util;
+
+/**
+ * Manipulate Docker Compose Files.
+ *
+ * @author Jesse Rushlow <jr@rushlow.dev>
+ *
+ * @internal
+ * @final
+ */
+class ComposeFileManipulator
+{
+    public const COMPOSE_FILE_VERSION = '3.7';
+
+    private $composeData;
+
+    public function __construct(string $contents)
+    {
+        if ('' === $contents) {
+            $this->composeData = $this->getBasicStructure();
+        } else {
+            $this->composeData = (new YamlSourceManipulator($contents))->getData();
+        }
+    }
+
+    public function getComposeData(): array
+    {
+        return $this->composeData;
+    }
+
+    public function serviceExists(string $name): bool
+    {
+        if (\array_key_exists('services', $this->composeData)) {
+            return \array_key_exists($name, $this->composeData['services']);
+        }
+
+        return false;
+    }
+
+    public function addDockerService(string $name, array $details): void
+    {
+        $this->composeData['services'][$name] = $details;
+    }
+
+    public function removeDockerService(string $name): void
+    {
+        unset($this->composeData['services'][$name]);
+    }
+
+    public function exposePorts(string $service, array $ports): void
+    {
+        $this->composeData['services'][$service]['ports'] = $ports;
```

Comment:
This is where we should add the comments we talked about:

```
$finalPortData = [];
foreach ($ports as $port) {
    $finalPortData[] = sprintf('__COMMENT__ exposes port %s to a random port on your host machine', $port),
    $finalPortData[] => sprintf('__COMMENT__ use %s:%s to expose to port %s on your host machine', $port, $port, $port),
    $finalPortData[] = $port;
}

// but, using the YSM in reality
$this->composeData['services'][$service]['ports'] = $finalPortData;
```

That code probably contains syntax errors - but hopefully you get the idea :). It's not optimized for multiple ports... because we don't have that use-case yet (but it would work fine).

Reply:
done